options/options.go

```diff
@@ -2213,16 +2216,38 @@ func (interp *interpreter) messageLiteralValue(mc *internal.MessageContext, fiel
 			if typeURLDescriptor == nil || typeURLDescriptor.Kind() != protoreflect.StringKind {
 				return interpretedFieldValue{}, reporter.Errorf(interp.nodeInfo(fieldNode.Name), "%vfailed to set type_url string field on Any: %w", mc, err)
 			}
-			msg.Set(typeURLDescriptor, protoreflect.ValueOfString(fullURL))
+			typeURLVal := protoreflect.ValueOfString(fullURL)
+			msg.Set(typeURLDescriptor, typeURLVal)
 			valueDescriptor := fmd.Fields().ByNumber(2)
 			if valueDescriptor == nil || valueDescriptor.Kind() != protoreflect.BytesKind {
 				return interpretedFieldValue{}, reporter.Errorf(interp.nodeInfo(fieldNode.Name), "%vfailed to set value bytes field on Any: %w", mc, err)
 			}
-			b, err := proto.MarshalOptions{Deterministic: true}.Marshal(msgVal.val.Message().Interface())
+
+			b, err := appendOptionBytes(nil, msgVal.msgVal)
```

Comment:
Thanks for mentioning this. It's also the case in the equals() implementation of protobuf-es (and I guess most if not all implementations based on the IEEE standard). It's surprising behavior when comparing messages, we should at least document it.

Reply:
Do I understand correctly that C++ marshals fields by field number ascending, but `proto.Marshal` marshals in order of occurrence in the source? 

I can't think of an easy way to work around that, especially considering Any. The serializer is compact and easy to read, so it doesn't seem like an issue to me, assuming we test it with all of internal/testdata/options, not just with options using Any.

Reply:
> Do I understand correctly that C++ marshals fields by field number ascending, but proto.Marshal marshals in order of occurrence in the source?

That is correct for the standard marshalling behavior. But `protoc` does something even stranger with the options message itself, and it is likely an artifact of how the dynamic message implementation works. It relies on the fact that binary format allows you to easily merge messages by concatenating their serialized form. So it actually serializes options this way:
1. All **known** fields are serialized in tag order.  This does not necessarily mean non-extension fields. If you provide a custom version of `descriptor.proto` that defines new non-extension fields for one of the options messages (not present in the version of `descriptor.proto` that was used to build the compiler), those are considered **unknown** fields. If you make incompatible changes in a custom version of `descriptor.proto` such that the wire type for a known field changes, it will interpret that data (known tag number but unexpected wire type) as an unknown field, too.
2. It then serializes all other (unknown) fields in the order in which they appear in source, as if each option statement were a separate value -- even if multiple options operate on the same value (via de-structuring a message value or repeated field value across multiple option statements).

There's more info including an example here: https://protobuf.com/docs/descriptors#encoding-options